src/app/modules/auth/auth.controller.ts

```diff
@@ -69,42 +70,60 @@ export const _handleLoginSendOtp: ControllerHandler<
     ...createReqMeta(req),
   }
 
-  return (
-    // Step 1: Validate email domain.
-    AuthService.validateEmailDomain(email)
-      // Step 2: Create login OTP.
-      .andThen(() => AuthService.createLoginOtp(email))
-      // Step 3: Send login OTP to email address.
-      .andThen((otp) =>
-        MailService.sendLoginOtp({
-          recipient: email,
-          otp,
-          ipAddress: requestIp,
-        }),
-      )
-      // Step 4a: Successfully sent login otp.
-      .map(() => {
-        logger.info({
-          message: 'Login OTP sent successfully',
-          meta: logMeta,
-        })
-
-        return res.status(StatusCodes.OK).json(`OTP sent to ${email}`)
+  // Step 1: Validate email domain.
+  const loginOtpResult = await AuthService.validateEmailDomain(email).andThen(
+    // Step 2: Create login OTP.
+    () => AuthService.createLoginOtp(email),
+  )
+  // Step 3a: Successfully created login otp.
+  if (loginOtpResult.isOk()) {
+    const { otp, otpPrefix } = loginOtpResult.value
+    return (
+      // Step 4: Send login OTP to email address.
+      MailService.sendLoginOtp({
+        recipient: email,
+        otpPrefix,
+        otp,
+        ipAddress: requestIp,
       })
-      // Step 4b: Error occurred whilst sending otp.
-      .mapErr((error) => {
-        logger.error({
-          message: 'Error sending login OTP',
-          meta: logMeta,
-          error,
+        // Step 5a: Successfully sent login otp.
+        .map(() => {
+          logger.info({
+            message: 'Login OTP sent successfully',
+            meta: logMeta,
+          })
+          return res
+            .status(StatusCodes.OK)
+            .json({ message: `OTP sent to ${email}`, otpPrefix })
```

Comment:
This constitute a change of response type (not just field within an envelop), so we must account for backward compatibility at deployment time, when new UI may hit old servers and vice-versa.

So just to confirm: it looks like the previous response was never used, so old UI hitting new server should be OK, correct? And for the case of new UI hitting old server, code would get `const { otpPrefix } = response` yielding `undefined` (but at least JS should not break with a runtime error, since it can still destructure a string), yes?

Site note: sad that the previous step 4a and 4b did not have the same response type (`string` and `{message:string}` respectively) 😢 , we should favour always having envelops, rather than returning primitives.

Reply:
Yep seems like what you've brought up is true! Will add the 2 cases into the PR 👍 
We can slowly improve on the the response types :')